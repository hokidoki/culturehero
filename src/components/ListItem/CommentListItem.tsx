import React from 'react';
import CommentListItemStyles from './CommentListItem.styles';
import { CommentsListItemProps } from './CommentListItem.types';
import CommentEditor from '../Editor/CommentEditor';
import { DashArrayCircleSpinnerWithLayout } from '../Spinner/DashArrayCircleSpinner';

export default function CommentListItem({
  createdAt,
  article,
  editing,
  id,
  setEdit,
  deleteComment,
  putComment,
  isLoading,
}: CommentsListItemProps) {
  if (editing)
    return (
      <CommentEditListItem
        isLoading={isLoading}
        putComment={putComment}
        deleteComment={deleteComment}
        id={id}
        article={article}
        createdAt={createdAt}
        setEdit={setEdit}
      />
    );

  return (
    <CommentViewListItem
      deleteComment={deleteComment}
      id={id}
      article={article}
      createdAt={createdAt}
      setEdit={setEdit}
    />
  );
}

interface CommentViewListItemProps
  extends Pick<
    CommentsListItemProps,
    'article' | 'createdAt' | 'setEdit' | 'deleteComment' | 'id' | 'isLoading'
  > {}

function CommentViewListItem({
  createdAt,
  setEdit,
  deleteComment,
  article,
  id,
  isLoading,
}: CommentViewListItemProps) {
  return (
    <CommentListItemStyles.Layout>
      {isLoading && <DashArrayCircleSpinnerWithLayout />}
      <CommentListItemStyles.ViewerHeader>
        <CommentListItemStyles.CreatedAt>
          {createdAt}
        </CommentListItemStyles.CreatedAt>
        <CommentListItemStyles.Controlls>
          <CommentListItemStyles.ControlButton
            color="plain"
            disabled={isLoading}
            onClick={() => setEdit(id)}
          >
            수정
          </CommentListItemStyles.ControlButton>
          <CommentListItemStyles.ControlButton
            color="red"
            disabled={isLoading}
            onClick={() => deleteComment(id)}
          >
            삭제
          </CommentListItemStyles.ControlButton>
        </CommentListItemStyles.Controlls>
      </CommentListItemStyles.ViewerHeader>
      <CommentListItemStyles.Article>{article}</CommentListItemStyles.Article>
    </CommentListItemStyles.Layout>
  );
}

interface CommentEditListItemProps
  extends Pick<
    CommentsListItemProps,
    | 'article'
    | 'createdAt'
    | 'setEdit'
    | 'deleteComment'
    | 'id'
    | 'putComment'
    | 'isLoading'
  > {}

function CommentEditListItem({
  article,
  createdAt,
  isLoading,
  id,
  setEdit,
  deleteComment,
  putComment,
}: CommentEditListItemProps) {
  return (
    <CommentListItemStyles.Layout>
      {isLoading && <DashArrayCircleSpinnerWithLayout />}
      <CommentListItemStyles.ViewerHeader>
        <CommentListItemStyles.CreatedAt>
          {createdAt}
        </CommentListItemStyles.CreatedAt>
        <CommentListItemStyles.Controlls>
          <CommentListItemStyles.ControlButton
            color="plain"
            disabled={isLoading}
            onClick={() => setEdit('')}
          >
            취소
          </CommentListItemStyles.ControlButton>
          <CommentListItemStyles.ControlButton
            color="red"
            disabled={isLoading}
            onClick={() => deleteComment(id)}
          >
            삭제
          </CommentListItemStyles.ControlButton>
        </CommentListItemStyles.Controlls>
      </CommentListItemStyles.ViewerHeader>
      <CommentEditor
        disabled={isLoading}
        article={article}
        onSubmit={(article) => putComment(article)}
      />
    </CommentListItemStyles.Layout>
  );
}
