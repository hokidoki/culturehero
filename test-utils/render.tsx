import 'jest-styled-components';
import { ThemeProvider } from 'styled-components';
import { render } from '@testing-library/react';
import theme from '../src/styles/Theme';
import { ReactNode } from 'react';

const Wrapper = ({ children }: { children: ReactNode }) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
);

const renderWithStyledComponent = (
  ui: JSX.Element,
  options?: Parameters<typeof render>[1]
) => render(ui, { wrapper: Wrapper, ...options });

export * from '@testing-library/react';

export { renderWithStyledComponent as render };
