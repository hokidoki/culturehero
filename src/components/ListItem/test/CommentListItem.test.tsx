import { fireEvent, render } from '@render';
import CommentListItem from '../CommentListItem';
import { act } from 'react-dom/test-utils';

describe('CommentListItem', () => {
  it('match snapshot when editing', () => {
    const props = {
      id: '0',
      article: 'Comment',
      createdAt: '22-02-22',
      editing: true,
      articleId: '0',
      setEdit: () => {},
      deleteComment: () => Promise.resolve(''),
      putComment: () => Promise.resolve(''),
      isLoading: false,
    };
    const utils = render(<CommentListItem {...props} />);
    expect(utils.container).toMatchSnapshot();
  });

  it('match snapshot when viewmode', () => {
    const props = {
      id: '0',
      article: 'Comment',
      createdAt: '22-02-22',
      editing: false,
      articleId: '0',
      setEdit: () => {},
      deleteComment: () => Promise.resolve(''),
      putComment: () => Promise.resolve(''),
      isLoading: false,
    };
    const utils = render(<CommentListItem {...props} />);
    expect(utils.container).toMatchSnapshot();
  });

  it('editing일때, 쓰기 버튼을 클릭하면 putComment가 호출된다.', async () => {
    const props = {
      id: '0',
      article: 'Comment',
      createdAt: '22-02-22',
      editing: true,
      articleId: '0',
      setEdit: () => {},
      deleteComment: () => Promise.resolve(''),
      putComment: jest.fn().mockResolvedValue(''),
      isLoading: false,
    };
    const utils = render(<CommentListItem {...props} />);
    const writeButton = utils.getByText('쓰기');

    await act(async () => fireEvent.click(writeButton));
    expect(props.putComment).toBeCalled();
  });

  it('삭제 버튼을 클릭하면 deleteComment가 호출된다.', async () => {
    const props = {
      id: '0',
      article: 'Comment',
      createdAt: '22-02-22',
      editing: true,
      articleId: '0',
      setEdit: () => {},
      deleteComment: jest.fn().mockResolvedValue(''),
      putComment: jest.fn().mockResolvedValue(''),
      isLoading: false,
    };
    const utils = render(<CommentListItem {...props} />);
    const writeButton = utils.getByText('삭제');

    await act(async () => fireEvent.click(writeButton));
    expect(props.deleteComment).toBeCalled();
  });

  it('취소 혹은 수정버튼을 클릭하면, setEdit가 호출된다.', async () => {
    const props = {
      id: '0',
      article: 'Comment',
      createdAt: '22-02-22',
      editing: true,
      articleId: '0',
      setEdit: jest.fn(),
      deleteComment: jest.fn().mockResolvedValue(''),
      putComment: jest.fn().mockResolvedValue(''),
      isLoading: false,
    };
    const utils = render(<CommentListItem {...props} />);
    const cancelbtn = utils.getByText('취소');

    await act(async () => fireEvent.click(cancelbtn));
    expect(props.setEdit).toBeCalledTimes(1);

    props.editing = false;
    utils.rerender(<CommentListItem {...props} />);
    const editBtn = utils.getByText('수정');
    await act(async () => fireEvent.click(editBtn));
    expect(props.setEdit).toBeCalledTimes(2);
  });
});
