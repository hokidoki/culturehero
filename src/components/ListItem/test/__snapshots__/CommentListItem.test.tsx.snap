// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentListItem match snapshot when editing 1`] = `
.c1 {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  height: 50px;
  width: 100%;
  padding: 0px 10px;
  border-bottom: 1px solid rgb(46,48,51);
}

.c2 {
  color: #bdc1c6;
  font-size: 1rem;
}

.c3 {
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  align-items: center;
  gap: 10px;
}

.c4 {
  background-color: transparent;
  border: none;
  color: #9aa0a6;
}

.c5 {
  background-color: transparent;
  border: none;
  color: palevioletred;
}

.c0 {
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
  position: relative;
  width: 100%;
  min-height: 200px;
  height: 200px;
  border: 2px solid rgb(46,48,51);
  border-radius: 10px;
}

.c6 {
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  align-items: flex-start;
  width: 100%;
  height: 100%;
  gap: 20px;
  flex-grow: 1;
  padding: 20px;
}

.c7 {
  width: 100%;
  height: 100%;
  border-radius: 10px;
  background-color: transparent;
  outline: none;
  resize: none;
  padding: 10px;
  font-size: 1rem;
  color: #9aa0a6;
  background-color: rgb(33,33,36);
  cursor: text;
}

.c7 div[contenteditable=true] {
  outline: none;
}

.c8 {
  width: 100px;
  height: 100%;
  border: 2px solid rgb(46,48,51);
  border-radius: 10px;
  background-color: transparent;
  font-size: 1.25rem;
  color: white;
  transition: all 0.5s;
}

.c8:disabled {
  color: #bdc1c6;
  cursor: not-allowed;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <span
        class="c2"
      >
        22-02-22
      </span>
      <div
        class="c3"
      >
        <button
          class="c4"
          color="plain"
        >
          취소
        </button>
        <button
          class="c5"
          color="red"
        >
          삭제
        </button>
      </div>
    </div>
    <div
      class="c6"
    >
      <div
        class="c7"
      >
        <div
          contenteditable="true"
        >
          Comment
        </div>
      </div>
      <button
        class="c8"
      >
        쓰기
      </button>
    </div>
  </div>
</div>
`;

exports[`CommentListItem match snapshot when viewmode 1`] = `
.c1 {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  height: 50px;
  width: 100%;
  padding: 0px 10px;
  border-bottom: 1px solid rgb(46,48,51);
}

.c2 {
  color: #bdc1c6;
  font-size: 1rem;
}

.c3 {
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  align-items: center;
  gap: 10px;
}

.c4 {
  background-color: transparent;
  border: none;
  color: #9aa0a6;
}

.c5 {
  background-color: transparent;
  border: none;
  color: palevioletred;
}

.c6 {
  padding: 10px;
  color: #9aa0a6;
  width: 100%;
  flex-grow: 1;
}

.c0 {
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
  position: relative;
  width: 100%;
  min-height: 200px;
  height: 200px;
  border: 2px solid rgb(46,48,51);
  border-radius: 10px;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <span
        class="c2"
      >
        22-02-22
      </span>
      <div
        class="c3"
      >
        <button
          class="c4"
          color="plain"
        >
          수정
        </button>
        <button
          class="c5"
          color="red"
        >
          삭제
        </button>
      </div>
    </div>
    <p
      class="c6"
    >
      Comment
    </p>
  </div>
</div>
`;
